### DISCUSSION TITLE
how to make widgets do not steal mouse events?

### DISCUSSION THREAD

**Discussion by Xtr4F** [Aylur/ags#515] [Q&A] [answered]
**Recently I wanted to make a widget that can be freely dragged around the screen with the mouse cursor. I used the code from here: https://github.com/Aylur/ags/discussions/303 and implemented roughly what I wanted:**

https://github.com/user-attachments/assets/f31adb52-1843-426b-b0d6-27784c23be49

_Full code_:  https://pastebin.com/80hLdFps

**As you can see from the video, when I hold down the mouse button and drag the widget everything works as it is supposed to, but when I try to interact with something outside the widget, nothing happens, as if `Widget.Fixed` steals the mouse clicks and won't let me click anything. So, the question is how do I fix this? How to make widgets do not "steal" mouse events?**

**Comment by Xtr4F** [comment]
Here's a clearer example

https://github.com/user-attachments/assets/f0eb2ffe-9cf6-4e45-aaad-a51f49c85505

**Comment by Aylur** [comment]
You have a full screen layer open, it's the same thing as if you had an application in full screen.
You can try setting `clickThrough` to false on the window

**Reply by Xtr4F** [reply]
When `clickThrough` is set to false nothing changes, when it is set to true I can actually interact with the underlying windows as i wanted, but my widget no longer receives any mouse event at all:

https://github.com/user-attachments/assets/6a99cefc-d3a8-46af-9ccc-7f5f881185df

How to make my widget still be able to receive mouse events?

**Reply by Aylur** [reply]
Yeah I meant setting it to true, my bad
You can't have it both ways since you have a fullscreen window, either only your window receives events or the windows below it

**Reply by kotontrion** [reply, marked as answer]
the clickThrough property will make the whole window not receive events. To achieve your goal you have to set the input region of the window to the match the the size/position of your widget. You have to do this manually, I do something similar to make the transparent parts of my bar clickthrough.

Here is a snippet on how you can do that:

```js
//import cairo
import Cairo from "cairo"

//do this after the the widget has been moved
const region = new Cairo.Region()
const alloc = event_box.get_allocation()
region.unionRectangle(alloc)
event_box.get_toplevel().input_shape_combine_region(region)
```

Note that `get_allocation()` returns the coordinates relative to the widgets parent, while input_region is relative to the window. In your example this does not matter, because the event_box parent has the same allocation as the window. If this were not the case you would have to translate the coordinates, which can be done by using gtks `translate_coordinates` function

**Reply by Xtr4F** [reply]
> the clickThrough property will make the whole window not receive events. To achieve your goal you have to set the input region of the window to the match the the size/position of your widget. You have to do this manually, I do something similar to make the transparent parts of my bar clickthrough.
> 
> Here is a snippet on how you can do that:
> 
> ```js
> //import cairo
> import Cairo from "cairo"
> 
> //do this after the the widget has been moved
> const region = new Cairo.Region()
> const alloc = event_box.get_allocation()
> region.unionRectangle(alloc)
> event_box.get_toplevel().input_shape_combine_region(region)
> ```
> 
> Note that `get_allocation()` returns the coordinates relative to the widgets parent, while input_region is relative to the window. In your example this does not matter, because the event_box parent has the same allocation as the window. If this were not the case you would have to translate the coordinates, which can be done by using gtks `translate_coordinates` function

It worked. Thank you so much!
